File: customer-java-webflux/src/main/java/com/prez/api/ElementNotFoundErrorHandler.java
package com.prez.api;

import static org.springframework.http.HttpStatus.NOT_FOUND;
import static org.springframework.web.reactive.function.server.ServerResponse.status;

import com.prez.api.dto.ErrorResponse;
import com.prez.exception.NotFoundException;
import org.springframework.web.reactive.function.server.HandlerFilterFunction;
import org.springframework.web.reactive.function.server.HandlerFunction;
import org.springframework.web.reactive.function.server.ServerRequest;
import org.springframework.web.reactive.function.server.ServerResponse;
import reactor.core.publisher.Mono;

public class ElementNotFoundErrorHandler implements HandlerFilterFunction<ServerResponse, ServerResponse> {
  @Override
  public Mono<ServerResponse> filter(ServerRequest request, HandlerFunction<ServerResponse> next) {
    return next
        .handle(request)
        .onErrorResume(NotFoundException.class, ex ->
            status(NOT_FOUND)
                .bodyValue(new ErrorResponse("NOT_FOUND", ex.getLocalizedMessage()))
        );
  }
}

File: customer-kotlin-webflux/src/main/kotlin/com/prez/api/ElementNotFoundErrorHandler.kt
package com.prez.api

import com.prez.api.dto.ErrorResponse
import com.prez.exception.NotFoundException
import org.springframework.http.HttpStatus.NOT_FOUND
import org.springframework.web.reactive.function.server.HandlerFilterFunction
import org.springframework.web.reactive.function.server.HandlerFunction
import org.springframework.web.reactive.function.server.ServerRequest
import org.springframework.web.reactive.function.server.ServerResponse
import reactor.core.publisher.Mono


class ElementNotFoundErrorHandler : HandlerFilterFunction<ServerResponse, ServerResponse> {
  override fun filter(request: ServerRequest, next: HandlerFunction<ServerResponse>): Mono<ServerResponse> {
    return next
        .handle(request)
        .onErrorResume(NotFoundException::class.java) { ex: NotFoundException ->
          ServerResponse.status(NOT_FOUND)
              .bodyValue(ErrorResponse("NOT_FOUND", ex.localizedMessage))
        }
  }
}